### PR TITLE
HPCC-17712 DOCS:Roxie SSL listener

### PR DESCRIPTION
**Pull request by g-pan**
Fix HPCC-17712 DOCS:Roxie SSL listener
Add documentation for enabling SSL security for Roxie

@JamesDeFabia  please review
@mckellyln  please review

Refer to the Installing and Running the HPCC System PDF built on the build box:
http://10.240.32.243/view/Docs/job/DocDev05/8/CE=ubuntu-14.04-amd64-1/artifact/build-docs/Release/docs/Installing_and_RunningTheHPCCPlatform-6.4.0-rc1.pdf   (apx. pages 90-93)
To see test build output and friendlier version of the doc. 

Signed-off-by: G-Pan <greg.panagiotatos@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

### CODE REVIEW

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (JamesDeFabia) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: Refer to comments we discussed offline for opening para.

Comment (JamesDeFabia) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: However, this is missing one important point. We need to stress that the certificate must be copied to EVERY Roxie node.

**Review by JamesDeFabia — CHANGES_REQUESTED** [review]
Comments inline

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (mckellyln) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: Unless you specify a full-path to the cert and key files in configmgr and that full path is shared among the Roxie nodes ...

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (mckellyln) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: I would add an 'a' so it reads there is a similar

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (JamesDeFabia) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: @mark, you mean a full path to a central location on one server to the cert and key? Wouldn't that cause a bottleneck if interrogated by every node at the same time?

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (mckellyln) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: from native to SSL

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (mckellyln) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: specified a full-path above this is the location ...

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (mckellyln) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: Again, unless you specified a full-path that was shared across nodes all cert and key files would need to be copied to all Roxie nodes.

**Comment by mckellyln** [comment]
@g-pan see comments in line.

**Comment by mckellyln** [comment]
Perhaps add that this is for the listener - so its for clients sending jobs to Roxie.  Roxie communicates with other Roxie nodes via udp and is NOT SSL secure.  This is only for sending jobs to Roxie over SSL instead of traditional non-SSL sockets.

**Comment by mckellyln** [comment]
Maybe also don't repeat the pushing out of the environment.xml part.  Could you maybe do this once after the ESP and Roxie listener SSL sections ?

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (mckellyln) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: I do not think 100 (or even 500) nodes all reading a small file over nfs at the same time would cause any significant delays.  No one here uses nfs, but I have seen it used elsewhere.

Comment (mckellyln) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: Perhaps its not worth mentioning about nfs.  It's just possible to specify _any_ full-path for the key and cert files.  And unless it is a shared path, the key and cert files need to be copied to all Roxie nodes.

Comment (mckellyln) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: If there is no full path then its assumed they live in /var/lib/HPCCSystems/<roxie-instance-name> dir.

**Comment by g-pan** [comment]
Several Changes made. @mckellyln  and @JamesDeFabia please review. Once approved I will squash and rebase.

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (JamesDeFabia) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: Including as a <chapter> would be better than as a <book>

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (JamesDeFabia) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: (image)??

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (JamesDeFabia) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: certificate filename

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (JamesDeFabia) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: key filename

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (JamesDeFabia) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: The default location for the certificate and key files is <emphasis>/var/lib/HPCCSystems/myroxie</emphasis>. You can specify a full path if you want these files in a different location. The certificate and key must <emphasis role="bold">must</emphasis> be available to every Roxie node.

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (JamesDeFabia) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: Is the image issue resolved?

**Review by JamesDeFabia — CHANGES_REQUESTED** [review]
Comments inline

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (g-pan) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: yes resolved: Please see most recent PDF file:  http://10.240.32.243/job/DocDevII/4/CE=ubuntu-14.04-amd64-1/

File: docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml
Comment (richardkchapman) [inline docs/Installing_and_RunningTheHPCCPlatform/Inst-Mods/ssl-esp.xml]: Is that a sensible default location? It's readable by all... @mckellyln

**Review by JamesDeFabia — APPROVED** [review]
I approve now, but defer to @richardkchapman regarding his comment about default location for files which may change software

**Comment by g-pan** [comment]
@mckellyln reviewed and approved source file offline. I may have squashed away the @richardkchapman comment, where he questioned the default location for the certificate and key files. (please see outdated commit). 
Otherwise this is fully approved and rebased ready for a clean merge.

**Comment by mckellyln** [comment]
@richardkchapman I cannot find your comment above - but the default location for the cert/key files in /var/lib/HPCCSystems/<roxie-instance-name> was chosen to be the same pattern as where ESP cert/key files are currently looked for.  These files need to be readable only by hpcc user and could be chmod'ed to protect them.
We could choose a different location such as a new dir:
/var/lib/HPCCSystems/certfiles
or a sub-dir such as:
/var/lib/HPCCSystems/<roxie-instance-name>/certfiles
or
/etc/HPCCSystems/certfiles
And those dirs would require same permission rules as the cert/key files.
The user can also specify any full path at creation time.

**Pull request merged**
